File: Leetcode_Code/98. Validate Binary Search Tree/98a.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode *root) {
        if (!root) return true;
        bool Boolean = true;
        isValidBST_dfs(root, Boolean);
        return Boolean;
    }

private:
    int last = INT_MIN;
    bool count = false;

    void isValidBST_dfs(TreeNode *root, bool &Boolean) {
        if (!root) return;
        isValidBST_dfs(root->left, Boolean);
        int current = root->val;
        if (current <= last && count) {
            Boolean = false;
        }
        count = true;
        // cout << current << " " << last << endl;
        last = current;
        isValidBST_dfs(root->right, Boolean);
    }
};

File: Leetcode_Code/98. Validate Binary Search Tree/98a.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    private TreeNode lastNode = null;
    private boolean bool = true;

    public boolean isValidBST(TreeNode root) {
        if (root == null) return true;
        isValidBST(root.left);
        if (lastNode != null && lastNode.val >= root.val) bool = false;
        else {
            lastNode = root;
            isValidBST(root.right);
        }
        return bool;
    }
}
